libsolidity/ast/Types.cpp

```diff
@@ -199,18 +148,78 @@ util::Result<TypePointers> transformParametersToExternal(TypePointers const& _pa
 	return transformed;
 }
 
+bool compareTypes(Type const* lhs, Type const* rhs)
+{
+	solAssert(lhs && rhs, "");
+	return lhs->richIdentifier() < rhs->richIdentifier();
 }
+using TypeSet = set<Type const*, decltype(compareTypes)*> ;
 
-vector<frontend::Type const*> solidity::frontend::oversizedSubtypes(frontend::Type const& _type)
+void storageOversizedSubtypesInner(
+	Type const* _type,
+	bool _includeType,
+	set<string>& io_seen,
+	TypeSet& io_oversizedSubtypes
+)
 {
-	set<StructDefinition const*> structsSeen;
-	TypeSet oversized;
-	oversizedSubtypesInner(_type, false, structsSeen, oversized);
-	vector<frontend::Type const*> res;
+	for (Type const* subtype: _type->subtypes())
+	{
+		string id = subtype->richIdentifier();
+		if (io_seen.count(id))
+			continue;
+		io_seen.insert(id);
+		bool includeType = (_type->category() != Type::Category::Struct);
+		storageOversizedSubtypesInner(subtype, includeType, io_seen, io_oversizedSubtypes);
+		io_seen.erase(id);
+	}
```

Comment:
Both `storageOversizedSubtypesInner()` and `validForLocation()` now use something like this to traverse the tree. This way you're doing more work than necessary because you encounter the same types multiple times and you also have to repeat this in every function that needs to walk the tree.

How about adding something like `subtypesRecursive()` that walks the tree and gives you a set without duplicates? Then each function could filter it (e.g. to remove structs like its done here) and do its thing on every type in the set. You could also add a parameter to specify the filter as a lambda.

Reply:
> Both storageOversizedSubtypesInner() and validForLocation() now use something like this to traverse the tree.

Yes. That's why I tried to place the functions next to each other.

> How about adding something like subtypesRecursive()

Do you mean a non-virtual `[Composite]Type::subtypesRecursive()` (or even a free function)? Yes, it makes sense to consider. (It would be more iterative than recursive though, something like the flood fill algorithm. And, BTW, `io_seen` would be gone.)

That would be another inefficiency though. When we search for collisions, we do full search. But then we search for storage validity, we are happy to stop at the first oversized type. But perhaps we do not care about performance here.

Reply:
> Do you mean a non-virtual [Composite]Type::subtypesRecursive() (or even a free function)? Yes, it makes sense to consider. (It would be more iterative than recursive though, something like the flood fill algorithm. And, BTW, io_seen would be gone.)

Yes, exactly. And it could be either recursive or iterative, both versions would be pretty simple, it's just the standard depth-first tree traversal. I'd tend towards the non-recursive approach myself exactly because it would not require passing mutable data through parameters or having to split it into two functions like in case of `storageOversizedSubtypesInner()`. But that's just my personal preference, both approaches are valid.

> That would be another inefficiency though. When we search for collisions, we do full search. But then we search for storage validity, we are happy to stop at the first oversized type. But perhaps we do not care about performance here.

I think that some extra overhead is fine if we're just reporting an error and exiting - that's because we're taking the hit just once. It matters more if we're not stopping because then we get the performance hit for every type that we check.

Also, `subtypesRecursive()` traverses the whole tree but it just collects the types and does not do anything heavy for each of them. It's `validForLocation()` that iterates over the returned items and runs the checks and it could still stop after the first item. In fact it would have more control because it would have all items available and could process them in any order.

One thing that might be useful in that case though would be ordering the items by their nesting level in the tree.

Reply:
I added `Type::allSubtypes()` and adjusted a bit the code around.

Reply:
I see that now it also includes the type itself. Which is good but i think it needs a better name. Maybe something like `flattenedSubtypeTree()`? A bit ugly and still does not unambiguously say that the current type is included but at least hints that it might be possible.

BTW, I like how much simpler the code became after refactoring out the recursive part. Always stating the type rather than sometimes type and sometimes variable is also a good idea.

Reply:
Yeah, it is usually a matter of definitions if a tree node is a subnode for itself. I renamed as suggested, and adjusted the documentation string. I did not document the order though, let it be an implementation detail :).

Regarding the terminology, I am not sure even about `subtype`. It can be read as `subclass`, i.e. a type derived from the given type.

Reply:
Yeah, `subtypes()` bothers me a bit too. How about `composedOf()` or `composition()`? With the second you could then call the recursive one `flattenedCompositionTree()`.

BTW, one thing to keep in mind for this design is whether it still makes sense after #909. Enums with data will essentially be unions so it probably does (with slight adjustment to size computation) but I just wanted to point it out here.

Reply:
This was covered by #9818 so I'm marking as resolved.